watcher/watcher.go

```diff
@@ -117,6 +123,138 @@ func (fetcher EthBurnLogFetcher) FetchBurnLogs(ctx context.Context, from uint64,
 	return resultChan, nil
 }
 
+type SolFetcher struct {
+	client           *solanaRPC.Client
+	gatewayStatePubk solanaSDK.PublicKey
+	gatewayAddress   string
+}
+
+func NewSolFetcher(client *solanaRPC.Client, gatewayAddress string) SolFetcher {
+	seeds := []byte("GatewayState")
+	programDerivedAddress := solana.ProgramDerivedAddress(pack.Bytes(seeds), multichain.Address(gatewayAddress))
+	programPubk, err := solanaSDK.PublicKeyFromBase58(string(programDerivedAddress))
+	if err != nil {
+		panic("invalid pubk")
+	}
+
+	return SolFetcher{
+		client:           client,
+		gatewayStatePubk: programPubk,
+		gatewayAddress:   gatewayAddress,
+	}
+}
+
+func (fetcher SolFetcher) FetchBurnLogs(ctx context.Context, from uint64, to uint64) (chan BurnLogResult, error) {
+	resultChan := make(chan BurnLogResult)
+
+	go func() {
+		func() {
+			defer close(resultChan)
+			for i := from; i < to; i++ {
+				nonce := i + 1
```

Comment:
Can you explain what this nonce is tracking?

Reply:
Added a comment in the code; it refers to the number of burns on the solana gateway (which starts indexing at 1)

Reply:
Ah, I assumed `i` was just the block number in this case.

Reply:
Nope. Solana scans over burn indexes, Ethereum scans block numbers, and then iterates over events (nonce is still technically the burn index)

Reply:
Makes sense.